Support extra config options for Sentry

Moress opened the pull request:
For now only dsn can be configured through the airflow.cfg. Need support 'http_proxy' option for example (it can't be configured through the environment variables). This change implements solution for supporting all existed (and maybe future) options for sentry configuration.

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

AbhiPrasad reviewed (commented):
This looks good to me in terms of the Sentry API, but will defer final approval to the maintainers. Even if some options are not possible (ex. adding a custom `transport`), the SDK should guard against it sufficiently, so it should be fine.

We should probably add a note to the docs (`airflow/docs/errors.rst`) that you can now supply additional configuration options based on the Python platform https://docs.sentry.io/error-reporting/configuration/?platform=python

Moress committed:
Support extra config options for Sentry

For now only dsn can be configured through the airflow.cfg. Need support 'http_proxy' option for example (it can't be configured through the environment variables). This change implements solution for supporting all existed (and maybe future) options for sentry configuration.

mik-laj commented on airflow/sentry.py:
```diff
@@ -81,13 +81,17 @@ def __init__(self):
 
-        dsn = conf.get("sentry", "sentry_dsn")
+        dsn = None
+        sentry_config_opts = conf.getsection("sentry") or {}
```
Can you add this option to reference documentation?
airflow/config_templates/config.yml
https://airflow.readthedocs.io/en/latest/configurations-ref.html

Moress replied:
This section has already existed in docs. I update section description in airflow/config_templates/config.yml with info of supporting additional configs from sentry docs.

mik-laj commented on airflow/sentry.py:
```diff
@@ -81,13 +81,17 @@ def __init__(self):
 
-        dsn = conf.get("sentry", "sentry_dsn")
+        dsn = None
+        sentry_config_opts = conf.getsection("sentry") or {}
```
getsection method returns `dict[str,str]`.  This means that options that are not strings will not work properly.
https://github.com/getsentry/sentry-python/blob/464ca8dda09155fcc43dfbb6fa09cf00313bf5b8/sentry_sdk/consts.py#L37
The following options will works:
```
ca_certs
dist
dsn
environment
http_proxy
https_proxy
ignore_errors
in_app_exclude
in_app_include
release
server_name
request_bodies
```
The following options will NOT works due to type mismatch:
```
attach_stacktrace
before_breadcrumb
before_send
debug
default_integrations
integrations
max_breadcrumbs
propagate_traces
sample_rate
send_default_pii
shutdown_timeout
transport
with_locals
```
This can cause very difficult to debug problems. It would be great if we supported more options or if an option is not supported, we would report a user-friendly error message.

Moress replied:
> getsection method returns dict[str,str]

It's not quite true. Here (https://github.com/apache/airflow/blob/master/airflow/configuration.py#L422) you can see that code is going through `_section` dict and trying to cast strings to proper type.
But fair enough it's not working for these props (list, dict and custom type is not supported by getsection):
```
integrations
in_app_include
in_app_exclude
ignore_errors
before_breadcrumb
before_send
transport
```
I will add some guard exception for these props.

AbhiPrasad commented on docs/errors.rst:
```diff
@@ -39,6 +39,8 @@ Add your ``SENTRY_DSN`` to your configuration file e.g. ``airflow.cfg`` under ``
 .. note::
     If this value is not provided, the SDK will try to read it from the ``SENTRY_DSN`` environment variable.
 
+You can supply `additional configuration options <https://docs.sentry.io/error-reporting/configuration/?platform=python>`__ based on the Python platform via [sentry] section.
```
Can you also quickly note the unsupported options here, just so everything is clear?

```
UNSUPPORTED_SENTRY_OPTIONS = frozenset(
        ("integrations", "in_app_include", "in_app_exclude", "ignore_errors",
         "before_breadcrumb", "before_send", "transport")
    )
```

Moress replied:
Done. Add this info in doc and config templates.

AbhiPrasad reviewed (approved):
Thanks for sticking with this for so long.

@mik-laj what else do is needed to get this PR through?
